lib/Backends/OpenCL/OpenCLDeviceManager.cpp

```diff
@@ -701,3 +731,105 @@ void OpenCLDeviceManager::runFunctionImpl(
+
+  tensor.moveToDevice((DeviceManager *)this, context);
+  return true;
+}
+
+/// Copies the device buffer associated with \p tensor to the host.
+/// The tensor must be resident on this device. If \p release is true, frees
+/// the device memory. Updates the tensor residency info.
+bool OpenCLDeviceManager::transferFromDevice(Tensor &tensor, bool release) {
+  // auto info = tensor.getDeviceResidencyInfo();
+  runtime::OpenCLDeviceTransferContext *ctx =
+      static_cast<runtime::OpenCLDeviceTransferContext *>(
+          tensor.getLocationContext());
+
+  OpenCLCommandQueue queue;
+  cl_command_queue_properties props =
+      clDoProfile ? CL_QUEUE_PROFILING_ENABLE : 0;
+  auto queueOrError = commandQueuePool_.requestCommandQueue(props);
+  CHECK(queueOrError) << "Unable to get command queue from pool.";
+  queue = std::move(queueOrError.get());
+
+  cl_int err = clEnqueueReadBuffer(queue.backingQueue, ctx->tensorBuffer,
+                                   /* blocking_read */ CL_FALSE, ctx->offset,
+                                   tensor.getUnpaddedSizeInBytes(),
+                                   tensor.getUnsafePtr(),
+                                   /* num_events_in_wait_list */ 0,
+                                   /* event_list */ nullptr,
+                                   /* event */ nullptr);
+  CHECK_EQ(err, CL_SUCCESS) << "Unable to copy data from the device";
+
+  clFinish(queue.backingQueue);
+  commandQueuePool_.returnCommandQueue(queue);
+
+  if (release) {
+    releaseDeviceTensor(tensor.getLocationContext());
+  }
+  tensor.clearDeviceResidency();
+  delete ctx;
+  return true;
+}
+
+bool OpenCLDeviceManager::transferTensorOnThisDevice(
+    Tensor &tensor, OpenCLDeviceTransferContext *sourceTransferContext,
+    OpenCLDeviceTransferContext *targetTransferContext, bool releaseBuffer) {
+  OpenCLCommandQueue queue;
+  cl_command_queue_properties props =
+      clDoProfile ? CL_QUEUE_PROFILING_ENABLE : 0;
+  auto queueOrError = commandQueuePool_.requestCommandQueue(props);
+  CHECK(queueOrError) << "Unable to get command queue from pool.";
+  queue = std::move(queueOrError.get());
+
+  cl_int err = clEnqueueCopyBuffer(
+      queue.backingQueue, sourceTransferContext->tensorBuffer,
+      targetTransferContext->tensorBuffer, sourceTransferContext->offset,
+      targetTransferContext->offset, tensor.getUnpaddedSizeInBytes(),
+      /* num_events_in_wait_list */ 0,
+      /* event_list */ nullptr,
+      /* event */ nullptr);
+  CHECK_EQ(err, CL_SUCCESS) << "Unable to copy data from the device";
+
+  clFinish(queue.backingQueue);
+  commandQueuePool_.returnCommandQueue(queue);
+
+  if (releaseBuffer) {
+    // release temporary cl_mem buffer
+    releaseDeviceTensor((void *)sourceTransferContext);
+  }
+  tensor.moveToDevice((DeviceManager *)this, targetTransferContext);
+  return true;
+}
+
+/// Releases the device buffer associated with \p tensor.
+bool OpenCLDeviceManager::releaseDeviceTensor(void *locationContext) {
```

Comment:
I think we spoke about it in person, but I think it would reduce error potential if this took in a Tensor and cleared the DeviceResidencyInfo. This way the caller has to always clear the residency info themselves.

Reply:
This is doable but not as simple as it sounds, I think. 
Writing my thoughts to consider the pros and cons:

DRI is then required to hold a pointer to its owning tensor.
This is a results of DRI'a destructor calling releaseDeviceTensor(Tensor &t). 
This was put in place to prevent device memory leaks, but can also be error prone as it requires the user to be aware of double-free flows. For easiest use, "releaseDeviceTensor" must be implemented to check for double-free. 

Technical details: moving DRI member init from definition to each Tensor constructor (adds "static" complexity). 
In addition, a weird handling will then be required for Tensor's move-constructor and assignment operator , which swaps DRI pointers.